lib/fieldhand/paginator.rb

```diff
@@ -86,6 +90,19 @@ def parse_response(query = {})
       response_parser
     end
 
+    def retryable_request(query)
```

Comment:
Should the retries be per `Paginator` instance or per _request_? E.g. if a single request fails too many times, should you be unable to use that `Paginator` again?

Perhaps instead the retry limit should be refreshed for every call to `parse_response`, e.g.

```ruby
def parse_response(query = {})
  response = retryable_request(query, retries) # here `retries` refers to the option
  # ...
end

def retryable_request(query, retries)
  response = request(query)
  raise ResponseError, response unless response.is_a?(::Net::HTTPSuccess)

  response
rescue ResponseError => e
  raise e unless retries > 0
  sleep(interval)

  retryable_request(query, retries - 1)
end
```

This does risk blowing the stack if the retry limit is too high though (as Ruby doesn't do automatic tail call optimisation yet)...

Reply:
This is really interesting. My main concern is that we are running each process in containers with as less memory as possible and that we do not have control on pages/records size.

Reply:
I suppose the key question is “should the retry limit be reset between pages?”

E.g. if the request for page 1 fails a few times, should that affect how much we retry fetching page 2?

My gut feeling is that it shouldn’t as the documentation suggests that `retries` is per HTTP request, not per `Paginator`.

What do you think?

If we want the retry limit is to be per request and we don’t want to risk blowing the stack, perhaps we could extract a simple `Requester` class that has a mutable `retries` instance variable we can decrement? That way, we could initialise a new `Requester` with the `retries` option on every request?